### PR TITLE
Add user interface for graph editor graph layout parameters

### LINKED ISSUE #133: Expose config of graph layout algorithm to users
**Is your feature request related to a problem? Please describe.**

The user cannot tweak the layout algorithm.

**Describe the solution you'd like**

On element of user interaction with the layout algorithm could be to expose https://github.com/hexatomic/hexatomic/blob/a8d368cf17427935455903d62c61406ea2e7f2ee/bundles/org.corpus_tools.hexatomic.graph/src/main/java/org/corpus_tools/hexatomic/graph/internal/SaltGraphLayout.java#L63 in the UI.

**Describe alternatives you've considered**

**Additional context**

Nothing urgent at all.

### PR DESCRIPTION
**Pull request by thomaskrause**
## Pull request type

 

Feature

## What is the current behavior?

Issue Number: closes #133 

## What is the new behavior?

- It is possible to configure the different margins in the graph editor view in the UI
  - Vertical margin between all nodes
  - Horizontal margin between the token nodes (affects also spacing between the ancestor nodes)
  - Additional margin between the token row and the other nodes, useful for adding space for depency edges 
- Restructured the user documentation for the graph editor because there is now more optional functionality that should not all be described on the first page.
- Moving (Swt)ResourceManager classes from the WindowBuilder into core bundle, since they are used by at least two bundles to manage the icons. Otherwise, we would have to duplicate these two classes. Since these classes are now in core, I adapted them to our code style and fixed multiple code smells. Also, on the different places where font and image resources are already created, the new manger classes are now used.

## Does this introduce a breaking change?

N/A

## Other information

Peek 2022-09-08 18-00

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added **or** the PR is neither a bug fix nor a feature
- [x] Docs have been reviewed and added / updated if needed **or** the PR is neither a bug fix nor a feature
- [x] The *Unreleased* section in CHANGELOG.md has been amended to reflect the changes in this PR if needed
- [x] Build (`mvn -Pcff verify`) was run locally and any changes were pushed
- [x] The pull request is against the correct branch (`main` for bug fixes, `develop` for new functionality)
- [x] [Dependencies and citation templates](https://github.com/hexatomic/hexatomic/tree/develop/releng/templates) have been updated where necessary **or** there are no new dependencies

---

# Checklist for maintainers

## Releases

- Do **NOT** push the green merge button on GitHub.  
Follow the latest version of the [developer/maintainer docs](https://hexatomic.github.io/hexatomic/dev/) for making releases.
- Check that license and citation information for dependencies are complete (each dependency has a folder in `/THIRD-PARTY/`).

### CODE REVIEW

File: docs/user/src/usage/graph-editor/README.md [hunk 1]
```diff
@@ -48,9 +37,5 @@ You can navigate the graph view as follows:
 
 - You can **center the view** around a specific point in the graph by double-clicking that point.
 
-If you don't like the layout of the graph, you can change it by simply dragging nodes with your mouse.
-
-## Console
+If you don't like the layout of the graph, you can change it by simply dragging nodes with your mouse or [adjust the layout parameters](./layout-params.md).
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/README.md:40]:
```suggestion
If you don't like the layout of the graph, you can change it by dragging nodes with your mouse or [adjusting the layout parameters](./layout-params.md).
```

Remove documentation smell, fix grammar.

File: docs/user/src/usage/graph-editor/README.md [hunk 2]
```diff
@@ -5,28 +5,17 @@ It provides a general visualization that displays all possible types of annotati
 
 ![Screenshot of the graph editor](graph-viewer.png)
 
-## Filters
+On the bottom of the graph editor, there is the console which you can use to actually edit the graph.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/README.md:8]:
```suggestion
On the bottom of the graph editor is the console that you can use to edit the graph.
```

Simply sentence.

File: docs/user/src/usage/graph-editor/filter.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Filter visible annotations
+
+To access additional filter options, click and expand a relevant category in the **Filter View** section.
+
+![Screenshot of the graph editor with expanded filter view](filter-view.png)
+
+You can choose to display spans and their annotations in the graph, by first expanding **Annotation Types** and the checking the checkbox **Spans**.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/filter.md:7]:
```suggestion
You can choose to display spans and their annotations in the graph, by first expanding **Annotation Types** and then checking the checkbox **Spans**. 
```

Fix typo.

File: docs/user/src/usage/graph-editor/filter.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Filter visible annotations
+
+To access additional filter options, click and expand a relevant category in the **Filter View** section.
+
+![Screenshot of the graph editor with expanded filter view](filter-view.png)
+
+You can choose to display spans and their annotations in the graph, by first expanding **Annotation Types** and the checking the checkbox **Spans**. 
+Spans are special nodes to collect a number of tokens and to annotate them all at once.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/filter.md:8]:
```suggestion
Spans are special nodes that "collect" a number of tokens.
Span annotations can then be made for that *specific set* of tokens, e.g., a phrase, clause or sentence.
```

Improve wording. "Annotating all at once" sounds like you would use spans to add annotations to all tokens, whereas spans in itself are meaningful units.

File: docs/user/src/usage/graph-editor/filter.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+If you want to learn more about spans, please read the [Salt documentation](http://corpus-tools.org/salt/#documentation).
+Similarly, you can show or hide pointing relations between nodes in the graph by using the checkbox **Pointing Relations**.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/filter.md:10]:
```suggestion

Similarly, you can show or hide pointing relations between nodes in the graph by using the checkbox **Pointing Relations**.
```

Add blank line to separate this point from the span explanation above.

File: docs/user/src/usage/graph-editor/filter.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+**Node Annotations** allows you to filter the segments that include annotations of a specific *name*.
+To add a filter criterion (also called a "facet"), search for an annotation name in the text field and select the matching annotation.
+This creates a box with the annotation name of the applied filter.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/filter.md:14]:
```suggestion
This creates a new filter badge with the annotation name of the applied filter.
```

I think "filter badge" is a better name for the thing.

File: docs/user/src/usage/graph-editor/filter.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+You can add more than one filter and any node, that contains any of the selected annotation names, will be shown.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/filter.md:15]:
```suggestion
You can add more than one filter. All nodes that contain any of the selected annotation names will be shown.
```

Separate sentences.

File: docs/user/src/usage/graph-editor/filter.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+Click on the close button to delete the annotation name filter again.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/filter.md:16]:
```suggestion
Click on the close button on a filter badge to delete the respective annotation name filter.
```

Make more precise.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
+# Change layout parameters
+
+Depending on the shown annotations and the specific graph, the default visualization of the graph might get cluttered.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:3]:
```suggestion
Depending on the visible annotations and the specific graph, the default visualization of the graph might be cluttered.
```

Improve wording.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+The graph editor will try to arrange the nodes using a default layout algorithm, which parameters you can adjust by expanding the "Graph Layout" section.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:4]:
```suggestion
The graph editor will try to arrange the nodes using a default layout algorithm.
You can adjust the layout parameters by expanding the "Graph Layout" section.
```

Separate sentences.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+![Graph editor view with expanded layout parameters](./layout-params.png)
+
+## Horizontal margin between token
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:8]:
```suggestion
## Horizontal margin between tokens
```

Use plural

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+![Horizontal margin parameter slider](./layout-param-horizontal-margin.png)
+
+
+There is some horizontal space between the token at the bottom of the graph.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:13]:
```suggestion
This parameter changes the horizontal space between the tokens at the bottom of the graph.
```

Improve wording.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+If the space is increased, this indirectly affects the space between other nodes, too.
+
+This is measured in "times of the average token width".
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:16]:
```suggestion
This parameter is measured in "multiples of the average token width".
```

Improve wording.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+So for \"0\" there is no margin, for "1" the margin has  the same width as the average token node, and for "2" the margin is twice as high as the average token node width.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:17]:
```suggestion
So for \"0\" there is no margin, for "1" the margin has the same width as the average token node, and for "2" the margin is twice as high as the average token node width.
```

Remove duplicate whitespace.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+![Effect of the settings 0, 1 and 2 for the horizontal margin parameter](horizontal-margin-examples.png)
+
+## Vertical margin between nodes
+
+![Vertical margin parameter slider](./layout-param-vertical-margin.png)
+
+Configures the vertical margin between *all* nodes.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:25]:
```suggestion
This parameter configures the vertical margin between *all* nodes.
```

Make this a sentence.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+This is measure in "times of the node height".
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:27]:
```suggestion
This is measured in "multiples of the node height".
```

Fix typo, wording.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+So for "0" there is no margin, for "1" the margin has the same height as the node.
+
+![Effect of the settings 0, 1 and 2 for the vertical margin parameter](vertical-margin-examples.png)
+
+
+## Vertical margin between token and non-token
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:33]:
```suggestion
## Vertical margin between tokens and non-tokens
```

Plural.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+![Vertical margin between token and non-token parameter slider](./layout-param-token-margin.png)
+
+
+Token are grouped in the lowest rank (at the bottom)
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:38]:
```suggestion
Tokens are grouped horizontally at the bottom of the graph visualization, i.e., in the bottom "row".
```

Introduce row before it is used later, improve wording.

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+To allow space for pointing relations, you can add a margin between the token row and the lowest annotation nodes.
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:39]:
```suggestion
To allow space for pointing relations, you can add a margin between the token "row" and the annotation nodes in the "row" just above it.
```

File: docs/user/src/usage/graph-editor/layout-params.md [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+This margin is in addition to the [vertical margin between nodes](#vertical-margin-between-nodes).
+
+A margin of "1" means there is one level left empty, an offset of "0"
```
Comment (sdruskat) [inline docs/user/src/usage/graph-editor/layout-params.md:42]:
```suggestion
A margin of "1" means there is one empty level added between the token "row" and the annotation node "row" above it, an offset of "0"
```

Make clearer.

**Review by sdruskat — APPROVED** [review]
Thanks! Great stuff.

I've added a number of wording documentation improvement suggestions. Otherwise this is good to go.

One thing I realized is that when changing horizontal spacing is that the current focus is lost, i.e., the view scrolls to a different section of the graph. Whether fixing this is in scope for this PR is up to you. Could be a new PR just as well (in which case feel free to turn this into an issue).

hex

**Comment by thomaskrause** [comment]
> Thanks! Great stuff.
> 
> I've added a number of wording documentation improvement suggestions. Otherwise this is good to go.
> 
> One thing I realized is that when changing horizontal spacing is that the current focus is lost, i.e., the view scrolls to a different section of the graph. Whether fixing this is in scope for this PR is up to you. Could be a new PR just as well (in which case feel free to turn this into an issue).
> 
> hex [ hex ](https://user-images.githubusercontent.com/3007126/191238834-c203a292-8691-41ec-b3a3-7ca3ccf4ca60.gif) [ ](https://user-images.githubusercontent.com/3007126/191238834-c203a292-8691-41ec-b3a3-7ca3ccf4ca60.gif)

I added an issue for this, since it seems to work under Linux with an example corpus for me (I have trouble creating GIF screenshots, but will add the example later). So this could be related to the specific example corpus or to the Windows platform.

**Pull request merged**